---
title: "TEanalysis-tutorial"
author: "Per Ludvik Brattaas"
date: "November 17, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## How to use the deseqTE object to analyse your TE read counts in deseqAbstraction

### 1. Set up environment

```{r set up envir, message=FALSE}
rm(list=ls())
gc()

## load packages
library(devtools)
library(DESeq2)
install_github("perllb/deseqAbstraction")
library(deseqAbstraction)
```

### 2. set path to you featureCount files 
#### - Path to directory where the featureCount output files are (both count table .txt and the summary file MUST be in the same path. 
#### - Important: The summary file needs to have identical name as the .txt file, with the .summary suffix

```{r set paths}
# set path and filenames

path <- "/home/pbrattaas/Dropbox (MN)/Per/PhD/Projects/DNAmeth/hNES DNMT1 KO/RNAseq/PairedParam/Quant/"
te.file <- "hg38.s2.unique.ERE.SENSE.notInNCBI.Exon.txt.test"
coding.file <- "hg38.s2.unique.NCBI.Exon.txt"
```

#### 3. Create new objects - one for NCBI quantifications and one for RepeatMasker annotations

```{r create ojects, message=FALSE}
# create object with coding/mrna/gene counts
coding <- deseqAbs$new("codingData",filename = paste(path,coding.file,sep = ""))
# create object with transpososn counts. here, genome (mm10 or hg38) of your analysis has to be defined! more genomes might be added soon
te <- deseqTE$new(name = "TE",filename = paste(path,te.file,sep=""),genome = "hg38")
```

#### 4. Assign colData data.frame. It MUST contain a "condition" column that describes the main conditions of your experiment. Also, it is good to add additional informal column, such as "sample" , "sampleID" etc. 

```{r}
# assign colData!
te$colData <- data.frame(samples=c("hNES_CTR_1","hNES_CTR_2","hNES_CTR_3","hNES_KO_1","hNES_KO_2","hNES_KO_3"),
                         condition=c(rep("CTR",3),rep("KO",3)))

coding$colData <- data.frame(samples=c("hNES_CTR_1","hNES_CTR_2","hNES_CTR_3","hNES_KO_1","hNES_KO_2","hNES_KO_3"),
                         condition=c(rep("CTR",3),rep("KO",3)))
```

#### 5. Mapping statics.
##### - percentage TEs? Plots each class
```{r} 
# check read and mapping stats

# how much of each TE class?
te$percentTE()

# how many reads are assigned (mapped) to the given annotation? not assigned = all reads mapping to the genome, but not overlapping the given annotation

# how many TE reads out of all mapped reads?
te$readsAssigned()
# how many reads map to the coding genome?
coding$readsAssigned()
```

#### 6. Run deseq, diffex, log2 normalization, rpkm conversion, average read on conditions
```{r}
# run 
te$fullAuto()
coding$fullAuto()
```

#### 7. PlotTE diffex maPlot, color significant TE entries
```{r}
te$maPlotTE(te$test$Default)

```

